src/lazy/streaming_raw_reader.rs

```diff
@@ -55,11 +57,34 @@ impl<Encoding: Decoder, Input: IonInput> StreamingRawReader<Encoding, Input> {
         }
     }
 
+    /// Gets a reference to the data source and tries to fill its buffer.
+    // This is a separate method to better encapsulate its use of `unsafe`
+    #[inline]
+    fn pull_more_data_from_source(&mut self) -> IonResult<usize> {
+        let input = unsafe { &mut *self.input.get() };
```

Comment:
Can you please document this use of unsafe? Specifically, why is the unsafe block required and how are we sure that this is, in fact, safe?